Enable non-MPI applications and get rid of controller / application handshake

cmcantalupo opened the pull request:
Relates to #2728
Relates to #2875

This draft PR differs from the issue-2728 branch, in that this preserves application feedback when MPI is disabled.

cmcantalupo committed:
Enable GEOPM HPC runtime to track non-MPI apps

- Add a GEOPM service API to track application PIDs for profiling
  + Application PIDs register a profile name with the service
  + Enables controller to track application lifetime through PIDs
  + With application lifetime tracking enabled independently from
    MPI, the GEOPM runtime can be used on non-MPI applications
    and MPI applicaitons that do not enforce GEOPM's CPU affinity
    requirements.
- This does not currently provide a mechanism to track application
  events (region ID, hints, or progress).
- Update active_sessions.schema.json to reflect new state held by geopmd
- Allow profiling to be disabled if GEOPM_TIMEOUT is set to -1
- Add library init and fini functions to enable and disable profiling.
- Switch ApplicationIO to use service to determine if application is up
  or down
- Add a geopm_pio_ interface to enable profile PID tracking
  + This interface will use the GEOPM Service DBus interfaces to track
    PIDs associated with profile names.
- Add DBus proxy implementation
- Add session open/close call for profiling
- Split injection of profiling APIs into separate shared object so
  controller doesn't request to be profiled when it loads libgeopm.so
- New library called libgeopmload will be used by the application
  to enable profiling.
- Preload libgeopmload in the launcher rather than libgeopm
- Do not preload geopmload for geopmctl

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add C and python interfaces to SharedMemory objects

- Switch profile shared memory creation to user mode

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Stop passing affinity masks over shared memory, use the OS

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Stop using status to get process affinity

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
!fixup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Start to purge ControlMessage and ProfileSampler

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Simplify Profile class

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
!fixup build errors in service directory

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix more compiler issues

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not unlink profile shared memory as user

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Delete ProfileTable

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup spec file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix shared memory creation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove sleep, wait is implemented in ApplictionIO.connect()

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move profile init into library init function

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Make sleep loop faster

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Enable controller to run without MPI

- Create a NullComm implementation for one rank only.
- Refactor build to generate libgeopm.so libgeopmload.so and geopmctl
  when --disable-mpi is provided
- Some general clean up of older code

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix build issues discovered while trying to build base RPM

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup issue with profile enablement

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
First draft of some integration tests for running controller without mpi

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Some debugging

- Epoch count is showing up in the trace.
- Region names are required for reporter to generate region data
  and these are missing (ApplicationIO just returns an empty set).
- Region hash and hint do not appear in the trace.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
First draft of README for non-MPI runtime integration tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a service interface to expose region names

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Delete region names after they are requested for a profile

- Probably want something more robust.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
fixup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Bring back the m_cpu_set to Profile class

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Slow down spin loop

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Test clean up

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Only append the host name to CSV files when MPI is enabled

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix a race by putting a sleep in the test

- This is a work around which makes the test function.
- Seems that if the controller creates the ApplicationRecordLog
  after the application has added records some records are lost.
- Needs a fix in the ApplicationSampler.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove error checking from example code to make it easier to read

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a test that uses the frequency map agent

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add some example output (minus trace files)

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix issue where reporter was not getting first or last update from record log

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Load the ProfileIOGroup with libgeopmload.so

Update the library constructor to inject the
ProfileIOGroup into PlatformIO whenever
libgeopmload.so is linked

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add GEOPM_NUM_PROC environment variable

- Determines the required number of PIDs to connect
  before controller starts.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update copyright

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup for GEOPM_NUM_PROC variable

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rename geopmpolicy package to geopm-runtime in spec file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix spec file names for tarball and directory

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez commented on service/geopmdpy/system_files.py:
```diff
@@ -723,6 +731,51 @@ def remove_batch_server(self, client_pid):
+            profile_name = self._sessions[client_pid].pop('profile_name')
+        except KeyError:
+            raise RuntimeError(f'Client PID {client_pid} requested to stop profiling, but it had not been started.')
+        # TODO: store region names in file in /run/geopm-service to enable clean restart
```
It's fine. Delete :)

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -284,51 +266,37 @@ namespace geopm
     }
 
-    void ApplicationSamplerImp::connect(const std::string &shm_key)
+    void ApplicationSamplerImp::connect(void)
```
Refactor (break this down into more functions, perhaps a helper function)

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -337,12 +305,32 @@ namespace geopm
+                for (int cpu_idx = 0; cpu_idx < m_num_cpu; ++cpu_idx) {
+                    if (CPU_ISSET_S(cpu_idx, set_size, cpuset.get())) {
+                        process.cpus.push_back(cpu_idx);
+                        // TODO: Last PID wins if the affinity masks overlap
```
Emit a warning that warns user about overlap

cmcantalupo committed:
Add session start stop to the library load functions

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Completely remove libgeopmpolicy

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup missing override

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup MockServiceProxy

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup many of the tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on src/ApplicationIO.cpp:
```diff
@@ -40,67 +56,83 @@ namespace geopm
+                // valid PID
+                break;
+            }
         }
```
Need a unit test to prove that a process that is erased from the middle of the vector does not lose the first elements.  E.g. a vector of 1,2,3 where 2 is erased will have pid_id point to 3.  What happened to 1?

bgeltz commented on src/Reporter.cpp:
```diff
@@ -53,7 +53,8 @@ namespace geopm
                       environment().policy(),
-                      environment().do_endpoint())
+                      environment().do_endpoint(),
+                      environment().timeout() != -1)
```
Reconsider if this is the best way for a user to toggle profiling on and off.  environment().do_profile() for example.

bgeltz commented on src/DefaultProfile.cpp:
```diff
@@ -10,40 +12,24 @@
-    {
-        g_pmpi_prof_enabled = 0;
-        g_pmpi_tprof_enabled = 0;
+        g_prof_enabled = 1;
```
Why are you enabling profiling in the destructor?  The seems to be the opposite of the previous behavior.

bgeltz commented on src/DefaultProfile.cpp:
```diff
@@ -10,40 +12,24 @@
-
-    DefaultProfile::~DefaultProfile()
-    {
-        g_pmpi_prof_enabled = 0;
-        g_pmpi_tprof_enabled = 0;
```
Why do we no longer need to set the enable vars to 0 in the destructor?

bgeltz commented:
Are there any guards against oversubscribing the cores between 2+ apps?

avilcheslopez commented on service/geopmdpy/system_files.py:
```diff
@@ -723,6 +731,51 @@ def remove_batch_server(self, client_pid):
+    def get_profile_region_names(self, profile_name):
+        result = None
+        if profile_name in self._region_names:
+            result = self._region_names.pop(profile_name)
```
Just curious: if (based on the method name) the intention is to retrieve the region names for a given profile, why pop the profile entry from `_region_names` altogether?

cmcantalupo replied:
We need some mechanism for clean up of the resources associated with the application run, and since we currently have only one consumer of this information (the controller), I put the clean up in the function the controller calls once at the end.  This is a nasty side effect of this method, and we should have a better story for cleaning up.

avilcheslopez replied:
So maybe it's just a matter of changing the method's name, for example, something like `pop_profile_region_names()`

bgeltz commented on src/ApplicationSampler.cpp:
```diff
@@ -98,7 +95,9 @@ namespace geopm
                                 environment().record_filter(),
-                                {})
+                                {},
+                                environment().timeout() != -1,
```
Another spot to revist if we want the timeout env var to control do_profile.

bgeltz commented on src/ApplicationStatus.hpp:
```diff
@@ -61,23 +61,7 @@ namespace geopm
-            /// @param [in] cpu_idx Index of the Linux logical CPU
-            /// @return ID of the process running on the given CPU.
-            virtual int get_process(int cpu_idx) const = 0;
+            virtual void set_valid_cpu(const std::set<int> &cpu_idx, bool is_valid) = 0;
```
Needs docu

avilcheslopez replied:
Please add code documentation for this method

bgeltz commented on src/Controller.cpp:
```diff
@@ -156,16 +156,18 @@ namespace geopm
                      Agent::num_policy(environment().agent()),
                      Agent::num_sample(environment().agent()),
-                     std::unique_ptr<TreeComm>(new TreeCommImp(ppn1_comm,
+                     geopm::make_unique<TreeCommImp>(
```
Do we still need to use the geopm flavor of make_unique here, or can this be moved over to the GCC implementation?  Maybe we should save it and do it all at once?

cmcantalupo replied:
can't hurt to use the std version in new code.

bgeltz commented on src/Controller.cpp:
```diff
@@ -323,22 +322,20 @@ namespace geopm
         m_tracer->update(m_trace_sample);
         m_profile_tracer->update(m_application_sampler.get_records());
-
+        m_reporter->update();
```
Do we need to call update() here and on line 338?

cmcantalupo replied:
no i think this is a rebase error.  The better place is below, just because it matches the sequence of calls in the step() function.

cmcantalupo commented on src/CSV.cpp:
```diff
@@ -29,9 +30,11 @@ namespace geopm
         , m_buffer_limit(buffer_size)
         , m_is_active(false)
     {
+#ifdef ENABLE_MPI
```
ENABLE_MPI comes automatically with the MPI m4 macros that we are using (which are not GEOPM sourced)
https://github.com/geopm/geopm/blob/dev/m4/lx_find_mpi.m4

avilcheslopez commented on src/Profile.cpp:
```diff
@@ -56,75 +55,33 @@ namespace geopm
-        , m_comm(comm)
         , m_curr_region_id(0)
         , m_current_hash(GEOPM_REGION_HASH_UNMARKED)
         , m_ctl_shmem(nullptr)
```
It seems like this property is no longer needed

avilcheslopez commented on src/Profile.cpp:
```diff
@@ -134,212 +91,74 @@ namespace geopm
+            tmp_msg[PATH_MAX-1] = '\0';
+            std::cerr << tmp_msg << std::endl;
         }
+        m_is_enabled = true;
```
Did you intend to make `m_is_enabled = true` even in the case where the prior init steps throw an exception?

avilcheslopez committed:
non-MPI app support: fix unit tests

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Non-MPI app support: fix build errors

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Non-MPI app support: fix test-dist

Add integration/test-no-mpi to dist

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo committed:
Avoid having the controller profile itself

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up shared memory keys when application shuts down without ending profiling

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Modify launcher to set GEOPM_NUM_PROC

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use versioned shared object file for preload

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Only remove shared memory keys for the control process

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Set GEOPM_NUM_PROC in the Launcher

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

avilcheslopez committed:
Conditionally disable MPI in geopmbench

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo committed:
Avoid open_session()/close_session() repeat calls

- Move start_profile() / stop_profile() into Profile object

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Print the controller run config

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Make lazy init even lazier

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix how EpochIOGroup is constructed

- Since the EpochIOGroup explicitly relies on the controller
  to gather data from the ApplicationRecordLog, it should only
  be registered in the memory context of the controller process.
- This also allows us to remove the lazy initialization requirement
  from the EpochIOGroup.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a slow loop check for application completion through Service API

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix configure option for enabling MPI

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add events for application begin and end

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update SampleAggregator to use application start time

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez committed:
Mock out MPI operations in ModelRegion classes

- Create a singleton class to manage geopmbench config
- When MPI is not enabled, simulate MPI operations in ModelRegion classes
  with sleep

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on src/Reporter.cpp:
```diff
@@ -110,6 +103,9 @@ namespace geopm
         if (m_do_profile && m_proc_region_agg == nullptr) {
             // ProcessRegionAggregator should not be constructed until
             // application connection is established.
+            init_sync_fields();
+            init_environment_signals();
```
Does this impact the report if profiling is disabled?  The update function is a noop in this case, so it seems like the generate function should also be a noop.

cmcantalupo committed:
Revert "Update SampleAggregator to use application start time"

This reverts commit 424142e9a1f6b73ccc0aa9744d5d465bce5e1408.

cmcantalupo committed:
Revert "Add events for application begin and end"

This reverts commit b86e217753e774a30621cc6abcfc7ab4269f6f71.

cmcantalupo committed:
Add enable/disable MPI define globally

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix way application sampler determines inactive cpus

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Remove example output from no-mpi tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd committed:
Fix mismatched mock for ApplicationStatus

Signed-off-by: Daniel Wilson <daniel1.wilson@intel.com>

dannosliwcd committed:
Add tests for the NullComm implementation

Signed-off-by: Daniel Wilson <daniel1.wilson@intel.com>

bgeltz committed:
Report error message in exception

- It was being printed in all cases previously.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add skip_unless_no_service_or_root decorator

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Fixup runtime unit tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Add skip decorator to test_geopmio

- Test runs iff the service is not running
  or the user is root.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Fix use of GEOPM_ENABLE_MPI in ctl main

- Also compilation fixes.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
fixup! Print the controller run config

cmcantalupo committed:
Modify test launcher to use application mode when GEOPMBENCH_NO_MPI is defined

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Never return invalid PIDs

- If a process fails to call stop_profile() prior
  to termination, the service may not remove the
  session for up to one second.
- In this case, do not return the PID as part of
  the get_profile_pids() API.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not profile lscpu call

- When creating a topo cache by a process that
  has LD_PRELOADED libgeopmload.so, do not create
  a fork bomb.
- Also avoids having lscpu be mistaken for an
  application process.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Promote PID mismatch from a logic error to a runtime error

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez committed:
Add first draft of integration test for multiple apps

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo committed:
Throw if application PID affinities overlap

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Enable retry if application affinities overlap

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Create Profile API to reset CPU affinity mask

- Call this method at the end of MPI_Init() to account for
  late calls to sched_setaffinity() by the MPI runtime.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rervert debug prints, also enable repeated calls to get affinity

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not let the controller track itself

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Retry CPU affinity calculation periodically

- Done on the application and controller side
- Controller has an exponential backoff algorithm when
  mask doesn't change

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a five second sleep to the beginning of all test apps

- With the asynchronous start of the controller and the appliction
  some data may be missed in the first second of app execution.
- Avoid making assertions about the first second of the run.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup initialization of EpochIOGroup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Account for invalid CPUs in EpochIOGroup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix test compile issues and some mock interface changes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add Scheduler class abstraction to enable mocking

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use Scheduler interface in ApplicationSampler and Profile

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Fix format string used in error message

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Fix up ApplicationSampler test mocking

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup mocking in ProfileTest

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
fixup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add unit tests for new ServiceProxy methods

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez committed:
Improvements for test_multi_app integration tests

- Fixes for "app launcher" (test_multi_app.sh) to correctly launch
  multiple apps and have them tracked
- Add tests with assertions to verify report reflects expectations

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz committed:
Add PERF_STATUS@core to test_frequency_map

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Remove unused events and add an event for affinity changes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add an API to insert the affinity event into the record log

- Add call when cpuset changes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use affinity event instead of back-off algorithm

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix issues found when debugging

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Switch ApplicationRecordLog record_s to use absolute time

- Fixup unit tests to handle new record_s type

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix shared memory region size to match new record structure

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not use m_proc_region_agg when it is invalid

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update to integration tests for application time

- Includes a simple sleep integration test

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use procfs to determine application process start time

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a record log event for starting profiling

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a record log event for stopping profiling

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Only load IOGroups without controls if not root

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move time zero calcluation until after overhead for starting GEOPM profile

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix GEOPM_TRACE_PROFILE feature to support new events

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add section to report for time spent in MPI_Init()

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Include init overhead in test_hint_time.py assertions

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update mock objects with new methods

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup unit tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update monitor integration test to include MPI_Init region in total

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix assertions in test_monitor.py integration test

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use sync-runtime in test_trace assertions

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix issue with MPI_Init region accounting

- Original implementation would call geopm_region_enter() prior
  to updating the CPU affinity.
- This would cause the ProfileIOGroup to report incorrect region hash
  per-cpu.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move MPI_Init entry/exit further out

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Debug commit

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove debug from last commit leaving fix for active cpus

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update the last hint when affinities change

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update status with current hint

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Tests pass!

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Revert test power balancer changes made for debug

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove debug prints and unneeded std::move

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Reset timeout

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup ordering in application sampler update and get related unit test passing

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix Profile unit tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Revert "Reset timeout"

This reverts commit c49302aca8397984b29d31cd63e9dac43674eae6.

cmcantalupo committed:
Do not pretend we sampled since the start of the app

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Revert "Revert "Reset timeout""

This reverts commit 2c586df712c243040e9250e6c3640f1b5c9d0c8b.

cmcantalupo committed:
Fixup affinity mapping

- Wait for 5 jiffies before trusting /proc/*/status
- Wait until all processes have reported they are
  affinitized

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Revert return type of client_pids()

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Pass affinity map from the app to the controller via multiple events

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Track GEOPM overhead time with an event

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on integration/test/test_multi_app.py:
```diff
@@ -0,0 +1,76 @@
+    def setUpClass(cls):
+        sys.stdout.write('(' + os.path.basename(__file__).split('.')[0] +
+                         '.' + cls.__name__ + ') ...')
+        subprocess.run(['/bin/bash', 'test_multi_app.sh'],
```
A simple solution to the PATH problem here is to assume that test_multi_app.sh script is in the same directory as the test_multi_app.py script.  The path to the script itself

```
os.path.join(os.path.dirname(os.path.abspath(__file__)), 'test_multi_app.sh') 
```

or something similar

cmcantalupo committed:
Fixup rebase issues

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove last update after event loop completes

- There should be data lost based on new implementation
- Avoids delayed last time stamp in trace which was breaking
  the sample rate test.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove exception when cpus are requested prior to affinity events

- In some cases requests for cpu map may be made prior to
  the application sampler receiving the affinity events
  through the record log.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup missing code for overhead events

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Cleanup test_hint_time trace file names

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove DBus call from the controller loop

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix formatting for overhead column in ProfileTracer

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not instantiate PlatformIO for every use of libgeopmload.so

- The original intent is to enable preloading libgeopmload.so to enable
  users of tools like geopmread to be able to read ProfileIOGroup signals.
- This use case is interesting, but the overhead is too high for multi-process
  applications which will never create a PlaformIO object otherwise.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Modify test_profile_overflow to use 1 proc per core

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup for test_monitor.py

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Cache signal and control names at geopmd startup

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove unneed json validation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up profile keys at geopmd startup

- Tracking profiling between restarts of geopmd is not supported.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Start in on documentation updates for non-MPI branch

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add blurb to README introduction

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Revert "Remove last update after event loop completes"

- This reverts commit 3669aaeca7c8bf46e42d0ca7ec61e2ef2f3b40a9.
- Modified to not update trace file with last sample
- Causes sample rate test to fail since last sample is
  not regular.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use a set in DomainNetMap since purpose is lookup

cmcantalupo committed:
Update tutorial README

- The build.sh script does not work cleanly without icc

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
fix unit tests due to clean-up changes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Extend service integration tests to cover all shmem segments

- Add start/stop profiling calls to python batch helper.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

avilcheslopez committed:
Make region count in report a float, adjust test_multi_app expectations

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz committed:
Add client uid and gid to active sessions

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Update geopm.7.rst to include GEOPM_NUM_PROC environment variable

- Remove MPI qualifier from applications that GEOPM supports in README.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
WIP - Update runtime web page

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get rid of libgeopmload.so, it is not needed.

- Simplify constructor function and add it to libgeopm.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove MPI requirement section from runtime.rst

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a connect method to profile to support forked processes

- Called by library constructor function

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Switch test_multi_app.py to use numactl, not srun

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Revert "Deprecate the --geopm-preload option for geopmlaunch"

- This reverts commit 4dee35bfaa104b746d2277d5407852423b1d76f1.
- We need to expose this option again because __constructor__
  library init function will execute with all created processes
  including shells and other processes that may not be accounted
  for with GEOPM_NUM_PROC.
- This allows the launcher to run scripts as well as MPI binary
  executables where the LD_PRELOAD option can be added to the script.
- When using MPI_Init() for start up, preloading would only
  impact MPI applications.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Ensure dlopen mock is used

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Fix tests that are not counting towards coverage

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Refactor TestActiveSessions

- Avoid unnecessary line breaks.
- Avoid double quotes.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add test for start/stop profile

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Fix issues in remove_client when profiling

- Remove redundant lines.
- Do not update the session file in remove_client.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add LD_PRELOAD to leftmost command in pipeline

- Resolves #2951.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Fixup rebase issues

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Work around for not using zes_init()

- To avoid having the geopmd process always persist when not calling zes_init()
  use geopmread/geopmwrite CLI to get signal and control name lists.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez commented on scripts/geopmpy/launcher.py:
```diff
@@ -396,6 +394,7 @@ def __init__(self, argv, num_rank=None, num_node=None, cpu_per_rank=None, timeou
         # Calculate derived values
         if self.rank_per_node is None and self.num_rank and self.num_node:
             self.rank_per_node = int_ceil_div(self.num_rank, self.num_node)
+        self.environ_ext['GEOPM_NUM_PROC'] = str(self.rank_per_node)
```
Does it matter that `self.rank_per_node` could be adjusted a bit further down in the code (line 439)?

avilcheslopez commented on scripts/geopmpy/launcher.py:
```diff
@@ -1133,7 +1148,7 @@ def preload_option(self):
         if self.config and self.config.get_preload():
             value = ':'.join((ll for ll in
-                              ('libgeopm.so', os.getenv('LD_PRELOAD'))
+                              ('libgeopm.so.1.0.0', os.getenv('LD_PRELOAD'))
```
I'm worried that I'm seeing this change multiple times. Is there a reason why each implementation of `Launcher` implements `preload_option()`? If it is necessary, I would recommend at least putting this string in a constant.

avilcheslopez commented on service/geopmdpy/service.py:
```diff
@@ -860,6 +938,22 @@ def PlatformReadSignal(self, signal_name, domain, domain_idx, **call_info):
+    def PlatformStopProfile(self, region_names, **call_info):
+        return self._platform.stop_profile(self._get_pid(**call_info), region_names)
+
+    # TODO: This method should check credentials
+    def PlatformGetProfilePids(self, profile_name):
+        return self._platform.get_profile_pids(profile_name)
+
+    # TODO: This method should check credentials
+    def PlatformGetProfileRegionNames(self, profile_name):
+        return self._platform.get_profile_region_names(profile_name)
```
Are these TODOs for this PR or for a later PR?

cmcantalupo replied:
Yes, lets open an issue.

cmcantalupo committed:
Define library as a launcher member

- Do no repeat libgeopm.so.1.0.0 in the launcher code.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on src/GEOPMBenchConfig.cpp:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    GEOPMBenchConfigImp::GEOPMBenchConfigImp()
+        : GEOPMBenchConfigImp(getenv("GEOPMBENCH_NO_MPI") == nullptr)
```
This should be included in the geopmbench.1 man page.

cmcantalupo replied:
Just pushed a patch.

cmcantalupo committed:
Add section to geopmbench man page about GEOPMBENCH_NO_MPI environment variable

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on integration/test/test_hint_time.cpp:
```diff
@@ -22,6 +23,7 @@ int main(int argc, char **argv)
     int comm_rank;
     MPI_Init(&argc, &argv);
     MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank);
+    sleep(5);
```
Is this still required?  If yes, we need to add some documentation somewhere that makes it clear that if you try to profile regions that occur too quickly after app startup, data will be lost.

cmcantalupo replied:
The data isn't "lost" we do not collect hardware samples until the controller is ready, but the app runs immediately.  For this reason, we cannot associate hardware metrics with the application data that *is* collected and processed (time stamped application data is not lost).  So, for tests that require hardware data to be collected in regions that occur early in the app, the sleep is required.

cmcantalupo committed:
Update version on main readme to 2.0.2

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on integration/test/test_hint_time.py:
```diff
@@ -153,9 +157,9 @@ def test_hint_time(self):
-            msg = "Application totals should have four seconds of total time"
-            expect = 4.0
-            actual = raw_totals['sync-runtime (s)']
+            msg = "Application totals should have nine seconds of total time"
+            expect = 9.0 + overhead_time
```
There should be a comment here explaining why the total time is different than what is setup in the test_hint_time.cpp (4 seconds presently).

cmcantalupo replied:
That is the extra 5 seconds of wait time at the beginning of geopmbench of course.

bgeltz commented on integration/test/test_hint_time.py:
```diff
@@ -99,9 +100,11 @@ def setUpClass(cls):
                                                         time_limit=time_limit)
             launcher.set_num_node(cls._num_node)
             launcher.set_num_rank(num_rank)
+            launcher.set_pmpi_ctl('application')
```
What prompted this change?

bgeltz replied:
```suggestion
```

bgeltz commented on integration/test/test_hint_time.py:
```diff
@@ -76,6 +76,7 @@ def setUpClass(cls):
                          '.' + cls.__name__ + ') ...')
         cls._test_name = 'hint_time'
         cls._report_path = 'test_{}.report'.format(cls._test_name)
+        cls._trace_path = f'test_{cls._test_name}.trace'
```
```suggestion
```
Traces are not used in this test.  Was this just for debug?

cmcantalupo replied:
Both changes may have been made to debug (application launch mode and creation of the trace data).

bgeltz commented on integration/test/test_hint_time.py:
```diff
@@ -99,9 +100,11 @@ def setUpClass(cls):
             launcher = geopm_test_launcher.TestLauncher(app_conf,
                                                         agent_conf,
                                                         cls._report_path,
+                                                        cls._trace_path,
```
```suggestion
```

bgeltz commented on integration/test/test_hint_time.py:
```diff
@@ -167,7 +171,7 @@ def test_hint_time(self):
             actual = raw_epoch['time-hint-memory (s)']
             util.assertNear(self, expect, actual, msg=msg)
-            msg = "Epoch should have three seconds of total time".format(region_name)
+            msg = "Epoch should have two seconds of total time".format(region_name)
```
Same thing here.  Somewhere there should be a comment explaining what to expect and why based on the code in the cpp.

bgeltz commented on src/geopmbench_main.cpp:
```diff
@@ -152,6 +157,7 @@ int main(int argc, char **argv)
             err = geopm_prof_exit(init_rid);
         }
         if (!err) {
+            sleep(5);
```
There should be a comment here, and probably also in the geopmbench.1 man page about why this is necessary.

bgeltz commented on integration/test/test_monitor.py:
```diff
@@ -96,21 +96,24 @@ def test_runtime(self):
 
     def test_runtime_epoch(self):
         '''Test that region and epoch total runtimes match.'''
+        initial_sleep_time = 5.0
```
Include a comment about where this magic number came from.  Also, perhaps this should be moved to a class member variable if it's used multiple times.

bgeltz commented on integration/test/test_multi_app.sh:
```diff
@@ -0,0 +1,34 @@
+export GEOPM_TRACE=${TEST_NAME}_trace.csv
+export GEOPM_TRACE_PROFILE=${TEST_NAME}_trace_profile.csv
+export GEOPM_PROFILE=${TEST_NAME}
+#export GEOPM_REPORT_SIGNALS=CPU_TIMESTAMP_COUNTER@package
```
```suggestion
```
Debug cruft?

bgeltz commented on integration/test/test_multi_app.sh:
```diff
@@ -0,0 +1,34 @@
+export GEOPM_NUM_PROC=2
+
+geopmctl &
+sleep 2
```
geombench has a sleep 5 in it.  Is this still needed?  If yes, please add a comment explaining why.  I'm guessing this is just a problem for stress-ng?

bgeltz replied:
Remove the sleep and re-test.

bgeltz commented on integration/test/test_omp_outer_loop.cpp:
```diff
@@ -21,6 +22,7 @@ int main(int argc, char **argv)
     int comm_rank = -1;
     int comm_size = -1;
     err = MPI_Init(&argc, &argv);
+    sleep(5);
```
Comment on why.  Please do this in all the places that have a newly added sleep 5.

bgeltz commented on integration/test/test_time.py:
```diff
@@ -0,0 +1,123 @@
+        host_names = self._report.host_names()
+        self.assertEqual(len(host_names), self._num_node)
+        for host in host_names:
+            found_nw = False
+            found_nw_mem = False
```
```suggestion
```
Unused

bgeltz commented on integration/test/test_time.py:
```diff
@@ -0,0 +1,123 @@
+            msg = "Application totals should have 10 seconds of total time"
+            expect = 10.0
+            actual = raw_totals['runtime (s)']
+            util.assertNear(self, expect, actual, msg=msg)
```
Why doesn't this test need to account for the extra magic 5 seconds that the others do?

bgeltz commented on integration/test-no-mpi/README.md:
```diff
@@ -0,0 +1,53 @@
+
+TODO: Add some assertions to your "tests".
+
+TODO: Add command line arguments to ``geopmctl`` that align with the
```
Have tracking issues been created for these TODOs?

avilcheslopez commented on service/geopmdpy/system_files.py:
```diff
@@ -728,6 +735,61 @@ def remove_batch_server(self, client_pid):
             sys.stderr.write(f'Warning: {write_fifo_path} file was left over, deleting it now.\n')
             os.unlink(write_fifo_path)
 
+    def start_profile(self, client_pid, profile_name):
+        profile_name = str(profile_name)
+        self.check_client_active(client_pid, 'start_profile')
+        uid, gid = self._pid_info(client_pid)
+        self._sessions[client_pid]['client_uid'] = int(uid)
+        self._sessions[client_pid]['client_gid'] = int(gid)
+        if len(self._profiles) == 0:
+            size = 64 * os.cpu_count()
+            shmem.create_prof('status', size, client_pid, uid, gid)
+        if profile_name in self._profiles:
+            self._profiles[profile_name].add(client_pid)
+        else:
+            self._profiles[profile_name] = {client_pid}
+        self._sessions[client_pid]['profile_name'] = profile_name
+        size = 57384
+        shmem.create_prof('record-log', size, client_pid, uid, gid)
+        self._update_session_file(client_pid)
```
There is a scenario that could cause problems with this code. Given how the client is going to call into these APIs, it should never happen, but it's still worth protecting against in case of a bug in the client or a malicious actor calling into the DBus APIs.

The scenario happens as a result of:
- A client can call `PlatformStartProfile` multiple times
- The code tracks the profile a client belongs to in two data structures: a map of profile names to PIDs (as a profile can have multiple PIDs), and the session data map (where a PID's profile name is stored)

The scenario is: a client calls `PlatformStartProfile` twice with two different profile names. The way the code currently is, it will add the client's PID to two different profiles in the profiles map, and would only track the last profile name with which it was called in the session data map. So when this client stops profiling, its profile will be removed from the session data map, but it will be removed only from one of the profiles in the profiles map.

I think the `shmem.create_prof()` call might fail in this scenario the second time its called and might prevent this scenario from happening, but I think it would be better to guard against it more explicitly.

avilcheslopez commented on service/geopmdpy/shmem.py:
```diff
@@ -0,0 +1,32 @@
+    result_cstr = gffi.gffi.new("char[]", name_max)
+    err = _dl.geopm_shmem_path_prof(shm_key_cstr, pid, uid, name_max, result_cstr)
+    if err < 0:
+        raise RuntimeError('geopm_shmem_destroy() failed: {}'.format(error.message(err)))
```
Did you mean 'geopm_shmem_path_prof() failed'?

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -1,23 +1,76 @@
+
+The GEOPM Runtime is software designed to enhance energy efficiency of
+applications though active hardware configuration.  The architecture is
+designed to provide an secure infrastructure to support a wide range
```
```suggestion
designed to provide a secure infrastructure to support a wide range
```

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -1,23 +1,76 @@
+The GEOPM Runtime is software designed to enhance energy efficiency of
+applications though active hardware configuration.  The architecture is
+designed to provide an secure infrastructure to support a wide range
+of tuning algorithms while solving three related challenges.
```
```suggestion
of tuning algorithms while solving three related challenges:
```

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -66,19 +119,18 @@ individual MPI application and enable the management of system power
+The GEOPM Runtime package provides the libgeopm shared object library.
+There are several command line tools included in GEOPM which have
+dedicated manual pages.  The :doc:`geopmlaunch(1) <geopmlaunch.1>`
+command line tool is used to launch an MPI application while enabling
```
```suggestion
command line tool is used to launch an application while enabling
```

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -108,14 +160,14 @@ Build Requirements
 build configure command line.
 
-The GEOPM HPC runtime requires support for MPI, the Message Passing
+The GEOPM Runtime requires support for MPI, the Message Passing
```
Where's the right spot to document the nuances about controller launch modes and MPI requirements?  The way this is, it's saying we require MPI in all cases.

bgeltz commented on service/geopmdpy/gffi.py:
```diff
@@ -34,25 +34,25 @@ def get_dl_geopmd():
 try:
-    _dl_geopmpolicy = gffi.dlopen('libgeopmpolicy.so.1',
+    _dl_geopm = gffi.dlopen('libgeopm.so.1',
                                   gffi.RTLD_GLOBAL|gffi.RTLD_LAZY)
```
```suggestion
    _dl_geopm = gffi.dlopen('libgeopm.so.1',
                            gffi.RTLD_GLOBAL|gffi.RTLD_LAZY)
```

avilcheslopez commented on service/src/IOGroup.cpp:
```diff
@@ -88,41 +88,42 @@ namespace geopm
+            register_plugin(NVMLIOGroup::plugin_name(),
+                            NVMLIOGroup::make_plugin);
 #endif
+        }
```
Minor suggestion to change the structure of the `if` statement to avoid possibly generating an empty true branch:

```
    if (getuid() == 0) {
        // register plugins that require root privilege
    }
#ifdef GEOPM_ENABLE_SYSTEMD
    else {
        // register ServiceIOGroup
    }
#endif
```

bgeltz commented on src/All2allModelRegion.cpp:
```diff
@@ -102,40 +110,52 @@ namespace geopm
+                ModelRegion::region_enter();
+                for (uint64_t i = 0; i < m_num_progress_updates; ++i) {
+                    ModelRegion::loop_enter(i);
+                    std::this_thread::sleep_for(std::chrono::microseconds(100));
```
geopmbench.1.rst should probably be updated to say that all2all becomes sleep when MPI is disabled.

We should document all the changed behaviors for the other regions as applicable.

bgeltz commented on src/ApplicationIO.cpp:
```diff
@@ -2,105 +2,139 @@
+            return result;
+        }
+        double timeout = m_timeout;
+        if (timeout < 0.0) {
+            timeout = 60.0;
         }
```
Is this just a paranoia check?

bgeltz commented on src/ApplicationRecordLog.cpp:
```diff
@@ -169,21 +158,90 @@ namespace geopm
+    void ApplicationRecordLogImp::affinity(const geopm_time_s &time, int cpu_idx)
+    {
+        std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
+        m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
```
I think we could consider a refactor here to use dynamic_cast instead of explicit C style casting.

bgeltz commented on tutorial/admin/README.md:
```diff
@@ -31,22 +31,22 @@ message will be displayed recommending corrective action or referring the user
 
 ## 1. GEOPM+SLURM static policy plugin
-GEOPM provides two libraries: `libgeopmpolicy` and `libgeopm`. `libgeopmpolicy`
+GEOPM provides two libraries: `libgeopm` and `libgeopm`. `libgeopm`
```
```suggestion
GEOPM provides two libraries: `libgeopm` and `libgeopmd`. `libgeopmd`
```

bgeltz commented on tutorial/admin/README.md:
```diff
@@ -31,22 +31,22 @@ message will be displayed recommending corrective action or referring the user
+The `libgeopm` library must be available in order to set up a system with
 the GEOPM+SLURM static policy plugin. The SLURM plugin requires
-`libgeopmpolicy` alone; it does not use `libgeopm` or MPI. If the GEOPM runtime
+`libgeopm` alone; it does not use `libgeopm` or MPI. If the GEOPM runtime
```
```suggestion
`libgeopmd` alone; it does not use `libgeopm` or MPI. If the GEOPM runtime
```

bgeltz commented on tutorial/admin/README.md:
```diff
@@ -31,22 +31,22 @@ message will be displayed recommending corrective action or referring the user
 capabilities](#2-geopm-runtime-capabilities) section in this guide.
 
-A.  To build `libgeopmpolicy` and a compatible `libgeopm_slurm.so`:
+A.  To build `libgeopm` and a compatible `libgeopm_slurm.so`:
```
```suggestion
A.  To build `libgeopmd` and a compatible `libgeopm_slurm.so`:
```

bgeltz replied:
We should revisit all of this docu.

bgeltz commented on integration/test-no-mpi/Makefile:
```diff
@@ -0,0 +1,12 @@
+#  Copyright (c) 2015 - 2023, Intel Corporation
```
Delete this entire directory.  It's redundant with the original integration tests when you disable MPI in geopmbench.

bgeltz commented on integration/test/test_hint_time.cpp:
```diff
@@ -22,6 +23,7 @@ int main(int argc, char **argv)
     int comm_rank;
     MPI_Init(&argc, &argv);
     MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank);
+    sleep(5);
```
Comment please on why this sleep is here.

bgeltz commented on src/Environment.cpp:
```diff
@@ -431,6 +437,11 @@ namespace geopm
         return std::stoi(lookup("GEOPM_TIMEOUT"));
```
```suggestion
```

bgeltz commented on src/ApplicationIO.cpp:
```diff
@@ -2,105 +2,139 @@
+        if (m_is_connected) {
+            return result;
+        }
+        double timeout = m_timeout;
+        if (timeout < 0.0) {
+            timeout = 60.0;
         }
```
```suggestion
```

bgeltz replied:
The Environment.cpp should also enforce that the timeout is positive.

cmcantalupo committed:
Raise an exception if user PID requests profiling twice

- This should not be possible if the DBus call is made in
  to support the DefaultProfile singleton object constructor.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rename GetProfileRegionNames DBus API to PopProfileRegionNames

- Also update all supporting API names.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get rid of unused member

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get rit of test-no-mpi directory

- It is replaced by geopm/tutorial/local and running
  the geopm/integration/test tests with GEOPMBENCH_NO_MPI enabled.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Small clean up to multi-app test

- Get rid of sleep after launching geopmctl, not needed
- Remove commented code

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Switch Controller.cpp to use std::make_unique

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get rid of debug changes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup error messages in geopm_shmem CFFI wrappers

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Invert logic to avoid empty if statement in some configs

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Delete dead code

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Some small doc and whitespace clean up

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update geopmbench man page with sleep documentation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix for test_hint_time.py, need to use application mode

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove launcher topo cache creation

- The service is now required for the GEOPM Runtime to
  start profiling, so we can also rely on the sevice
  to provide the topo cache.
- Some systems may not have a /tmp mounted, or the user
  may not have permission to write.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez commented on src/ApplicationIO.hpp:
```diff
@@ -27,10 +30,10 @@ namespace geopm
             virtual ~ApplicationIO() = default;
             /// @brief Connect to the application via
             ///        shared memory.
```
Minor note: maybe add a note about what the function now returns.

avilcheslopez commented on src/ApplicationIO.hpp:
```diff
@@ -39,34 +42,34 @@ namespace geopm
             bool m_is_connected;
-            ApplicationSampler &m_application_sampler;
+            std::shared_ptr<ServiceProxy> m_service_proxy;
+            const std::string m_profile_name;
+            const std::string m_report_name;
+            const int m_timeout;
```
Nit: should these be uppercase?

cmcantalupo replied:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es9-avoid-all_caps-names

I have never been a fan of constant members being all caps, although I am not in the majority wrt contributors to the project.  Consistency is important, but there are a number of arguments against all caps members.

cmcantalupo committed:
Add comment about magic number and related github issue

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez commented on src/ApplicationRecordLog.hpp:
```diff
@@ -183,22 +166,28 @@ namespace geopm
         protected:
             ApplicationRecordLog() = default;
-            static constexpr size_t M_LAYOUT_SIZE = 49192;
+            static constexpr size_t M_LAYOUT_SIZE = 57384;
```
It's out of scope for this review, but still wanted to ask: why does the interface contain implementation details? Is it due to the static methods just above? If so, do those need to be static?

avilcheslopez commented on src/ApplicationSampler.hpp:
```diff
@@ -119,37 +103,19 @@ namespace geopm
-            /// used by the application that is connected.
-            /// The default shared memory key consists of the prefix
-            /// "/geopm-shm-" concatenated with the euid.
+            /// string as a key.
```
This code documentation corresponded to the `connect()` method that was deleted. The new connect method and other new methods need are missing their code documentation.

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -220,6 +201,83 @@ namespace geopm
+        m_status->update_cache();
+        double time_delta;
+        if (m_is_first_update) {
+            for (int cpu_idx = 0; cpu_idx != m_num_cpu; ++cpu_idx) {
```
```suggestion
            for (int cpu_idx = 0; cpu_idx < m_num_cpu; ++cpu_idx) {
```

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -220,6 +201,83 @@ namespace geopm
+        }
+        else {
+            time_delta = geopm_time_diff(&m_update_time, &curr_time);
+            for (int cpu_idx = 0; cpu_idx != m_num_cpu; ++cpu_idx) {
```
```suggestion
            for (int cpu_idx = 0; cpu_idx < m_num_cpu; ++cpu_idx) {
```

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -236,27 +294,28 @@ namespace geopm
     {
+        uint64_t result = GEOPM_REGION_HINT_INACTIVE;
         if (!m_status) {
-            return GEOPM_REGION_HINT_INACTIVE;
+            result = GEOPM_REGION_HINT_INACTIVE;
+        }
```
Could simply delete this `if` statement, if the intention is to return `GEOPM_REGION_HINT_INACTIVE` two cases:
- `!m_status`
- `m_status && !m_is_cpu_active`

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -282,67 +341,61 @@ namespace geopm
+                ++num_active_cpu;
             }
         }
+        for (int cpu_idx = 0; cpu_idx != m_num_cpu; ++cpu_idx) {
```
```suggestion
        for (int cpu_idx = 0; cpu_idx < m_num_cpu; ++cpu_idx) {
```

avilcheslopez commented on src/ApplicationSampler.cpp:
```diff
@@ -353,6 +406,58 @@ namespace geopm
+        return result;
+    }
+
+    std::vector<int> ApplicationSamplerImp::client_pids(void) const
```
Does this need to return a vector? Could it return a set, instead?
